### PR TITLE
Add RPC extensibility feature to dev

### LINKED ISSUE #9291: Map WebJobs RPC extensions onto the gRPC host
This issue is to track the work related to consuming the new WebJobs extensions that will be added as part of https://github.com/Azure/azure-webjobs-sdk/pull/2978. The necessary work for the functions host is to:

When the WebJobs script host is ready:
1. discover `WebJobsRpcEndpointDataSource` from the new script host's `IServiceProvider`.
2. map them onto the gRPC hosts available endpoints.

Additionally, as part of mapping the endpoints we should:

1. Ensure the gRPC service is resolved from a new scope of the script hosts `IServiceProvider`
2. Replace the `HttpContext.RequestServices` with the script hosts `IServiceProvider`.

### PR DESCRIPTION
**Pull request by jviau**
### Issue describing the changes in this PR

See https://github.com/Azure/azure-functions-host/pull/9292

resolves #9291

### Pull request checklist

* [ ] My changes **do not** require documentation changes
    * [ ] Otherwise: Documentation issue linked to PR
* [ ] My changes **should not** be added to the release notes for the next release
    * [x] Otherwise: I've added my notes to `release_notes.md`
* [x] My changes **do not** need to be backported to a previous version
    * [ ] Otherwise: Backport tracked by issue/PR #issue_or_pr
* [x] My changes **do not** require diagnostic events changes
    * Otherwise: I have added/updated all related diagnostic events and their documentation (Documentation issue linked to PR)
* [x] I have added all required tests (Unit tests, E2E tests)
    * Modified tests, did not add any as this is out-of-proc scenario only.

### CODE REVIEW

File: src/WebJobs.Script.Grpc/Server/ExtensionsCompositeEndpointDataSource.cs [hunk 1]
```diff
@@ -0,0 +1,300 @@
...
+namespace Microsoft.Azure.WebJobs.Script.Grpc
+{
+    /// <summary>
+    /// Endpoint data source which composes all the WebJobs extension data sources together.
+    /// </summary>
+    /// <remarks>
+    /// Implementation is adapted from <see cref="CompositeEndpointDataSource"/>.
+    /// https://github.com/dotnet/aspnetcore/blob/main/src/Http/Routing/src/CompositeEndpointDataSource.cs.
+    /// </remarks>
+    internal sealed class ExtensionsCompositeEndpointDataSource : EndpointDataSource, IDisposable
```
Comment (liliankasem) [inline src/WebJobs.Script.Grpc/Server/ExtensionsCompositeEndpointDataSource.cs:24]: Is it possible to write any unit or integration tests for this?

**Review by liliankasem — APPROVED** [review]
lgtm; just wondering if there's any additional tests we can write for the new bits added

File: src/WebJobs.Script.Grpc/Server/ExtensionsCompositeEndpointDataSource.cs (hunk 1 above)
Comment (jviau) [reply src/WebJobs.Script.Grpc/Server/ExtensionsCompositeEndpointDataSource.cs:24]: Sure, I'll add some soon.

Comment (jviau) [reply src/WebJobs.Script.Grpc/Server/ExtensionsCompositeEndpointDataSource.cs:24]: Added unit tests

File: src/WebJobs.Script.Grpc/Server/ExtensionsCompositeEndpointDataSource.cs [hunk 2]
```diff
@@ -0,0 +1,300 @@
...
+
+        /// <inheritdoc />
+        public void Dispose()
+        {
+            _scriptHostManager.ActiveHostChanged -= OnHostChanged;
+
+            List<IDisposable> disposables = null;
+            lock (_lock)
+            {
+                _disposed = true;
```
Comment (fabiocav) [inline src/WebJobs.Script.Grpc/Server/ExtensionsCompositeEndpointDataSource.cs:74]: We should be able to return here if disposed, correct?

Multiple threads entering here will go through the through the same list initialization and disposal flow as the lock is released.

**Review by fabiocav — APPROVED** [review]
A couple of small comments. Main focus was on the dependency updates, and those look good.

**Pull request merged**